pkg/manifests/external_dns.go

```diff
@@ -54,41 +86,86 @@ func (p Provider) String() string {
 	}
 }
 
-func (p Provider) ResourceName() string {
-	switch p {
-	case PublicProvider:
-		return externalDnsResourceName
-	case PrivateProvider:
-		return externalDnsResourceName + "-private"
-	default:
-		return ""
-	}
+// ExternalDnsConfig contains externaldns resources based on input configuration
+type ExternalDnsConfig struct {
+	//internally exposed
+	tenantId, subscription, resourceGroup,
+	clientId, serviceAccountName, namespace,
+	crdName string
+	identityType  IdentityType
+	resourceTypes []ResourceType
+	provider      provider
+
+	// externally exposed
+	resources          []client.Object
+	labels             map[string]string
+	dnsZoneResourceIDs []string
 }
 
-func (p Provider) Labels() map[string]string {
-	labels := map[string]string{
-		k8sNameKey: p.ResourceName(),
-	}
-	return labels
+func (e *ExternalDnsConfig) Resources() []client.Object {
+	return e.resources
 }
 
-// ExternalDnsConfig defines configuration options for required resources for external dns
-type ExternalDnsConfig struct {
-	TenantId, Subscription, ResourceGroup string
-	Provider                              Provider
-	DnsZoneResourceIDs                    []string
+func (e *ExternalDnsConfig) Labels() map[string]string {
+	return e.labels
 }
 
-// ExternalDnsResources returns Kubernetes objects required for external dns
-func ExternalDnsResources(conf *config.Config, externalDnsConfigs []*ExternalDnsConfig) []client.Object {
-	var objs []client.Object
+func (e *ExternalDnsConfig) DnsZoneResourceIds() []string {
+	return e.dnsZoneResourceIDs
+}
+
+func NewExternalDNSConfig(conf *config.Config, tenantId, subscription, resourceGroup, clientId, serviceAccountName, namespace, crdName string, identityType IdentityType, resourceTypes []ResourceType, provider provider, dnszoneresourceids []string) *ExternalDnsConfig {
```

Comment:
the point of having a constructor is validating some input. we should validate input and return an error here so we can make assumptions about the ExternalDnsConfig

Reply:
added validation